cpp/src/fil/infer.cu

```diff
@@ -198,8 +201,46 @@ struct finalize_block {
     : tmp_storage(tmp_storage_), num_classes(num_classes_) {}
 
   template <int NITEMS>
-  static size_t smem_footprint() {
-    return sizeof(typename BlockReduceHostBestClass<NITEMS>::TempStorage);
+  static size_t smem_footprint(bool predict_proba) {
+    return predict_proba
+             ? sizeof(typename BlockReduceHost<NITEMS>::TempStorage)
+             : sizeof(typename BlockReduceHostBestClass<NITEMS>::TempStorage);
+  }
+
+  template <int NITEMS>
+  __device__ __forceinline__ vec<NITEMS, float> shifted_exp(
+    vec<NITEMS, float> margin, vec<NITEMS, float> max) {
+    vec<NITEMS, float> res;
+    for (int i = 0; i < NITEMS; i++) res[i] = expf(margin[i] - max[i]);
+    return res;
+  }
+
+  template <int NITEMS>
+  __device__ __forceinline__ vec<NITEMS, float> block_all_max(
```

Comment:
`block_all_max()` and `block_all_sum()` differ only in the reduction operation. Consider having a reduction template parameterized by the reduction operation instead.